File: 84.largest-rectangle-in-histogram.cpp
/*
 * @lc app=leetcode id=84 lang=cpp
 *
 * [84] Largest Rectangle in Histogram
 */
#include <bits/stdc++.h>

using namespace std;

// @lc code=start
struct SegementTree {
    vector<int> st;
    int n;

    SegementTree(vector<int>& arr) {
        this->n = arr.size();
        this->st = vector<int>(4*n, 0);
        build(arr, 0, n-1, 1);
    }

    void build(vector<int>& arr, int L, int R, int idx) {
        if (L > R) { return; }
        if (L == R) { st[idx] = arr[L]; }
        else {
            int mid = (L + R) / 2;
            build(arr, L, mid, 2*idx);
            build(arr, mid+1, R, 2*idx+1);
            st[idx] = min(st[2*idx], st[2*idx+1]);
        }
    }

    int query(int query_l, int query_r) { return query(query_l, query_r, 0, n-1, 1); }
    int query(int query_l, int query_r, int L, int R, int idx) {
        if (query_l > query_r) { return INT_MAX; }
        if (L==query_l && R==query_r) { return st[idx]; }
        int mid = (L + R) / 2;
        int left_val = query(query_l, min(query_r, mid), L, mid, 2*idx);
        int right_val = query(max(query_l, mid+1), query_r, mid+1, R, 2*idx+1);
        return min(left_val, right_val);
    }
};

class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        //return segementTree(heights);
    }

    int segmentTree(vector<int>& heights) { // TLE
        int n = heights.size();
        SegementTree st(heights);
        int max_area = 0;
        for (int i=0; i<n; i++) {
            for (int j=i; j<n; j++) {
                int min_height = st.query(i, j);
                max_area = max(min_height*(j-i+1), max_area);
            }
        }
        return max_area;
    }

    int stackBased(vector<int>& heights) {
        int n = heights.size();
        stack<int> st;
        
    }
};
// @lc code=end

File: 84.largest-rectangle-in-histogram.py
#
# @lc app=leetcode id=84 lang=python3
#
# [84] Largest Rectangle in Histogram
#
from typing import List

# @lc code=start
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        # For every height[i],
        # 1. Find the next smallest bar to the left of heights[i].
        # 2. Find the next smallest bar to the right of heights[i].
        # 3. For each heights[i], the left next smallest and the right 
        # next smallest bar indicies determine the largest rectangle 
        # that can be created that has its width equal to heights[i].

        n = len(heights)
        stack = []

        leftmost = [-1] * n
        for i in range(n):
            while stack and heights[stack[-1]] >= heights[i]:
                stack.pop()
            if stack:
                leftmost[i] = stack[-1]
            stack.append(i)

        stack.clear()
        rightmost = [n] * n
        for i in range(n-1, -1, -1):
            while stack and heights[stack[-1]] >= heights[i]:
                stack.pop()
            if stack:
                rightmost[i] = stack[-1]
            stack.append(i)

        maxArea = 0
        for i in range(n):
            leftBoundary = leftmost[i] + 1
            rightBoundary = rightmost[i] - 1
            length = rightBoundary - leftBoundary + 1
            width = heights[i]
            area = length * width
            maxArea = max(area, maxArea)
        
        return maxArea
# @lc code=end

